google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/startup/AppEngineStandardUnsupportedJavaVersionCheck.java

```diff
@@ -76,29 +74,26 @@ public void runActivity(@NotNull Project project) {
     ApplicationManager.getApplication()
         .runReadAction(
             () -> {
-              for (Module module : projectModules) {
-                AppEngineStandardFacet appEngineFacet =
-                    AppEngineStandardFacet.getAppEngineFacetByModule(module);
-                if (appEngineFacet != null) {
-                  // this is a standard app
-                  if (!appEngineFacet.isNonStandardCompatEnvironment()) {
-                    // this is targeting the standard environment
-                    if (!appEngineFacet.isJava8Runtime()) {
-                      // The runtime only supports Java 7 or below.
-                      if (usesJava8OrGreater(module)) {
-                        invalidModules.add(module);
-                      }
-                    }
-                  }
-                }
-              }
+              Arrays.stream(projectModules)
+                  .filter(this::usesJava8OrGreater)
+                  .filter(
+                      module -> {
+                        AppEngineStandardFacet facet =
+                            AppEngineStandardFacet.getAppEngineFacetByModule(module);
+                        return !(facet == null || facet.isNonStandardCompatEnvironment())
+                            && facet
+                                .getRuntimeJavaVersion()
```

Comment:
this bit confuses me a bit.  Why should getRuntimeJavaVersion return a stream to be filtered?

Reply:
It's not a stream, it's an `Optional`, which provides the `.filter()` method to further filter the `Optional`, potentially turning it to `.empty()` if the filter doesn't match the value.